convert-hf-to-gguf.py

```diff
@@ -132,12 +172,25 @@ def set_gguf_parameters(self):
         self.gguf_writer.add_file_type(self.ftype)
         print(f"gguf: file type = {self.ftype}")
 
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        del bid  # unused
+
+        return [(self.map_tensor_name(name), data_torch)]
+
+    def extra_f32_tensors(self, name: str, new_name: str, bid: int | None, n_dims: int) -> bool:
+        del name, new_name, bid, n_dims  # unused
+
+        return False
+
+    def extra_f16_tensors(self, name: str, new_name: str, bid: int | None, n_dims: int) -> bool:
+        del name, new_name, bid, n_dims  # unused
+
+        return False
+
     def write_tensors(self):
-        block_count = self.hparams.get("n_layers", self.hparams.get("num_hidden_layers", self.hparams.get("n_layer")))
-        tensor_map = gguf.get_tensor_name_map(self.model_arch, block_count)
-        for name, data_torch in self.get_tensors():
+        for name, data_torch in self.tensors.items():
             # we don't need these
-            if name.endswith((".attention.masked_bias", ".attention.bias", ".attention.rotary_emb.inv_freq")):
+            if name.endswith((".attention.masked_bias", ".attention.bias", ".rotary_emb.inv_freq")):
```

Comment:
If you want to blacklist tensors in a per model basis you can use 
https://github.com/ggerganov/llama.cpp/blob/6ecf3189e00a1e8e737a78b6d10e1d7006e050a2/gguf-py/gguf/constants.py#L732-L743

Reply:
Nice idea! I didn't know this was there. Right now it's only used for `MODEL_ARCH.LLAMA` in `convert.py`. The exclusion lists for the other architectures were never really tested (e.g. `MODEL_ARCH.PLAMO` and `MODEL_ARCH.GROK` both have `MODEL_TENSOR.ATTN_ROT_EMBD` in `MODEL_TENSORS`, but they are not in `MODEL_TENSOR_SKIP`, since they rely on matching the name like you pointed out).

I'm not sure about using this; `modify_tensors` can already be used to skip tensors, and I think it's cleaner this way.

https://github.com/ggerganov/llama.cpp/blob/3e5e0dced5dfa2ce1a997767d3ecfeea86a79aeb/convert-hf-to-gguf.py#L1970-L1972

It was already done similarly in `write_tensors` (I tried to keep most of the model-specific code from `write_tensors` unchanged into `modify_tensors`). Otherwise, this means tensor types that *won't* be serialized should *still* get a serialized name, even if they're never used. BERT's pooling layers don't have an enum, and they probably should not have one if it's to be carried around but still remain unused like `LLM_TENSOR_ATTN_ROT_EMBD` in the `llama.cpp` file. (if they do get used, I agree they should get added, but they aren't for now.)

https://github.com/ggerganov/llama.cpp/blob/a2ac89d6efb41b535778bfeaecaae8fe295b6ed3/llama.cpp#L488

So instead, maybe it would be worth considering removing `MODEL_TENSOR.ROPE_FREQS` and `MODEL_TENSOR.ATTN_ROT_EMBD`? But this is likely out of the scope of this PR.